exist-core/src/main/java/org/exist/xquery/functions/fn/transform/TreeUtils.java

```diff
@@ -64,20 +65,41 @@ static StringBuilder pathTo(final Node node) {
     static List<Integer> treeIndex(final Node node) {
         final Node parent = node.getParentNode();
         if (parent == null) {
-            return new ArrayList<>();
+          final List<Integer> index = new ArrayList<>();
+          // The root element always index 0 within the document node.
+          // Some node implementations (e.g., org.exist.dom.memtree.NodeImpl) do not always have an associated document.
+          // In this case, the nodeIndex must get an extra 0 index to be valid for xdmDocument.
+          if (! (node instanceof Document)) {
+            index.add(0);
+          }
+          return index;
         }
         final List<Integer> index = treeIndex(parent);
-        Node sibling = node.getPreviousSibling();
+        Node sibling = previousSiblingNotAttribute(node);
         int position = 0;
         while (sibling != null) {
             position += 1;
-            sibling = sibling.getPreviousSibling();
+            sibling = previousSiblingNotAttribute(sibling);
         }
         index.add(position);
 
         return index;
     }
 
+    /**
+     * A org.exist.dom.persistent.StoredNode returns attributes of an element as previous siblings of the element's children.
+     * This is not compatible with the way xdmNodeAtIndex works, so we need to compensate for this.
+     * @param node
+     * @return the previous sibling of `node` that is not an attribute.
+     */
+    private static Node previousSiblingNotAttribute(Node node) {
+      Node sibling = node.getPreviousSibling();
+      if (sibling instanceof Attr) {
+        return null;
+      }
+      return sibling;
+    }
+
     static XdmNode xdmNodeAtIndex(final XdmNode xdmNode, final List<Integer> index) {
```

Comment:
Should this method not be adapted? This loops over `children()`, are these not including attributes?